tensorzero-core/src/providers/groq.rs

```diff
@@ -724,20 +778,52 @@ async fn tensorzero_to_groq_assistant_messages(
                     message: "Tool results are not supported in assistant messages".to_string(),
                 }));
             }
-            ContentBlock::File(file) => {
-                let resolved_file = file.resolve().await?;
-                let ObjectStorageFile { file, data } = &*resolved_file;
-                if file.detail.is_some() {
-                    tracing::warn!(
-                        "The image detail parameter is not supported by Groq. The `detail` field will be ignored."
+            ContentBlock::File(file) => match file.as_ref() {
+                LazyFile::Url {
+                    file_url:
+                        FileUrl {
+                            mime_type,
+                            url,
+                            detail,
+                        },
+                    future: _,
+                } if !fetch_and_encode_input_files_before_inference
+                    && matches!(
+                        mime_type.as_ref().map(mime::MediaType::type_),
+                        Some(mime::IMAGE) | None
+                    ) =>
+                {
+                    if detail.is_some() {
+                        tracing::warn!(
+                                "The image detail parameter is not supported by Groq. The `detail` field will be ignored."
+                            );
+                    }
+                    warn_cannot_forward_url_if_missing_mime_type(
+                        file,
+                        fetch_and_encode_input_files_before_inference,
+                        PROVIDER_TYPE,
                     );
+                    assistant_content_blocks.push(GroqContentBlock::ImageUrl {
+                        image_url: GroqImageUrl {
+                            url: url.to_string(),
+                        },
+                    });
                 }
-                assistant_content_blocks.push(GroqContentBlock::ImageUrl {
-                    image_url: GroqImageUrl {
-                        url: format!("data:{};base64,{}", file.mime_type, data),
-                    },
-                });
-            }
+                _ => {
```

Comment:
would be nice to be able to unit test this but I understand the fetching is inherently hard. 
Did you think at all about ways to consolidate this logic among providers? Not obvious to me how you would do it.

Reply:
Tough with custom types